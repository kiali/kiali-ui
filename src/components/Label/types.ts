export interface Labels {
  name: string;
  value: string;
}
