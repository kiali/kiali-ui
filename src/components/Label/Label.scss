.label-collection {
  overflow: hidden;
}

.label-pair {
  margin: 3px 2px 2px;
  max-width: 100%;
  float: left;
  display: block;

  .label {
    min-width: 50px;

    &:first-child {
      border-radius: 2px 0 0 2px;
    }

    &:last-child {
      border-radius: 0 2px 2px 0;
    }
  }

  .label-key {
    background-color: #bee1f4;
    color: #00659c;
    padding-top: 4px;
    border-radius: 2px 0 0 2px;
    padding-bottom: 4px;
    margin-left: 0;
  }

  .label-value {
    background-color: #7dc3e8;
    color: #fff;
    padding-top: 4px;
    border-radius: 0 2px 2px 0;
    padding-bottom: 4px;
    margin-left: 0;
  }
}
