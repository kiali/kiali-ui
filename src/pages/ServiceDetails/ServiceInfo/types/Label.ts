export default class Label {
  key: string;
  value: string;

  constructor(key: string, value: string) {
    this.key = key;
    this.value = value;
  }

  isValid() {
    return typeof this.key === 'string' && typeof this.value === 'string';
  }
}
