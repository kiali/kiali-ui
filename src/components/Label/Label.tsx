import * as React from 'react';
import { Label as PfLabel } from 'patternfly-react';
import './Label.css';
import { canRender } from '../../utils/SafeRender';

interface Props {
  name: string;
  value: string;
}

const Label = (props: Props) => {
  const { name, value } = props;

  if (canRender(name) && canRender(value)) {
    return (
      <span className="label-pair">
        <PfLabel bsStyle="primary" className="label-key">
          {name}
        </PfLabel>
        <PfLabel bsStyle="primary" className="label-value">
          {value || ''}
        </PfLabel>
      </span>
    );
  } else {
    return <span>This label has an unexpected format</span>;
  }
};

export default Label;
