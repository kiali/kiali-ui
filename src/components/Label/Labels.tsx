import * as React from 'react';
import Label from './Label';
import { style } from 'typestyle';

const SHOW_MORE_TRESHOLD = 2;

interface Props {
  labels?: { [key: string]: string };
}

interface State {
  expanded: boolean;
}

const linkStyle = style({
  float: 'left',
  margin: '7px 2px 2px 3px',
  fontSize: '0.8rem'
});

class Labels extends React.Component<Props, State> {
  constructor(props: Props, state: State) {
    super(props, state);
    this.state = {
      expanded: false
    };
  }

  labelKeys() {
    return Object.keys(this.props.labels || {});
  }

  hasLabels() {
    return this.labelKeys().length > 0;
  }

  hasManyLabels() {
    return this.labelKeys().length > SHOW_MORE_TRESHOLD;
  }

  showItem(i: number) {
    return this.state.expanded || !this.hasManyLabels() || i < SHOW_MORE_TRESHOLD;
  }

  expandLabels = () => {
    this.setState({ expanded: true });
  };

  renderMoreLabelsLink() {
    if (this.hasManyLabels() && !this.state.expanded) {
      return (
        <a key="label_more" className={linkStyle} onClick={this.expandLabels}>
          {' '}
          More labels...
        </a>
      );
    }

    return null;
  }

  renderLabels() {
    return this.labelKeys().map((key, i) => {
      const hideClass = this.showItem(i) ? '' : 'hide';
      return (
        <div key={'label_' + i} className={hideClass}>
          <Label key={'label_' + i} name={key} value={this.props.labels ? this.props.labels[key] : ''} />
        </div>
      );
    });
  }

  renderEmptyLabels() {
    return <span> No labels </span>;
  }

  render() {
    if (this.hasLabels()) {
      return [this.renderLabels(), this.renderMoreLabelsLink()];
    } else {
      return this.renderEmptyLabels();
    }
  }
}

export default Labels;
