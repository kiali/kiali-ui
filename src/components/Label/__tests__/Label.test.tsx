import * as React from 'react';
import { shallow } from 'enzyme';
import Label from '../Label';
import { shallowToJson } from 'enzyme-to-json';

const mockBadge = (name = 'my_key', value = 'my_value') => {
  const component = <Label value={value} name={name} />;
  return shallow(component);
};

describe('#Badge render correctly with data', () => {
  it('should render badge', () => {
    const key = 'app';
    const value = 'bookinfo';
    const wrapper = mockBadge(key, value);

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();

    expect(wrapper.name()).toEqual('span');
    expect(wrapper.props().className).toEqual('label-pair');

    const labelKey = wrapper.find('Label').getElements()[0];
    const labelValue = wrapper.find('Label').getElements()[1];

    expect(labelKey.props.className).toEqual('label-key');
    expect(labelValue.props.className).toEqual('label-value');

    expect(labelKey.props.children).toEqual(key);
    expect(labelValue.props.children).toEqual(value);
  });
});
