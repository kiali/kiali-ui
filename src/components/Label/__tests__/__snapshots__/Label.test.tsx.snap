// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Badge render correctly with data should render badge 1`] = `
<span
  className="label-pair"
>
  <Label
    bsStyle="primary"
    className="label-key"
    type="default"
  >
    app
  </Label>
  <Label
    bsStyle="primary"
    className="label-value"
    type="default"
  >
    bookinfo
  </Label>
</span>
`;
